[MM-32927] Upgrade chart.js from version 2.9.4 to 3.8.2

m1lt0n opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
Upgrade chart.js dependency from 2.9.4 to 3.8.2 in the context of regular dependency upgrades in the web app.

There are some breaking changes that are addressed in the places where chart.js is used (according to the [migration guide](https://www.chartjs.org/docs/latest/getting-started/v3-migration.html) of the library.

Note: As far as I've checked, the charts are used in top channels insights and in system console. Checking manually the components seem to behave the same as before.

<!--
A description of what this pull request does.
-->

### Testing
Channel insights need a professional or enterprise plan to be visible.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-34915
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Upgrade chart.js from version "2.9.4" to "3.8.2"
```

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

m1lt0n commented on NOTICE.txt:
```diff
@@ -322,7 +322,7 @@ Simple HTML5 charts using the canvas element.
 The MIT License (MIT)
 
-Copyright (c) 2018 Chart.js Contributors
+Copyright (c) 2014-2022 Chart.js Contributors
```
Synced the licence file of the library (minor change).

M-ZubairAhmed commented on components/analytics/doughnut_chart.tsx:
```diff
@@ -4,7 +4,8 @@
 import React from 'react';
 import {FormattedMessage} from 'react-intl';
-import Chart, {ChartData} from 'chart.js';
+import Chart from 'chart.js/auto';
```
wondering 🤔 why we need `/auto`

m1lt0n replied:
The new version require either registering via:

```
Chart.register(...registerables); // registerables are imported from chart.js as well
```

and auto does that (importing and registering the relevant controllers / line/doughnut etc).

m1lt0n commented on package.json:
```diff
@@ -20,7 +20,7 @@
     "bootstrap": "3.4.1",
     "buffer": "6.0.3",
-    "chart.js": "2.9.4",
+    "chart.js": "3.8.2",
```
Switching to 3.8.2 instead of the latest 3.9.1, as the latter has a bug related to rollup that causes type declarations to not be found, which makes type checks fail.

nevyangelova reviewed (approved):
LGTM! Running e2e-test to double check all use cases

nevyangelova commented:
/e2e-test

nevyangelova commented:
@m1lt0n some e2e tests are failing but doesn't seem related to me https://mattermost-cypress-report.s3.amazonaws.com/250369-e0123d9-onprem-ent-webapp-pr-11326-mm-ee-test:e0123d9/mochawesome.html

can you double check just in case?

m1lt0n commented:
/update-branch

m1lt0n commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/250429

m1lt0n commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/250824

m1lt0n commented:
^ @nevyangelova the latest e2e run succeeded (without any changes to the code, since failures were either in unrelated code or there were errors pulling images in the workers).

BenCookie95 commented:
A few issues with insights:
1. There is a legend appearing above the graph that we don't currently have
2. The graph is no longer responsive. If you look at master, the graph resizes when you dynamically change page width but this one doesn't seem to do so. 

See screenshot for both issues
![Screen Shot 2022-10-14 at 8 51 16 AM](https://user-images.githubusercontent.com/14115924/195851611-32c56fec-88a6-4d74-8b8e-977c618bb13a.png)

BenCookie95 reviewed (changes requested):
A few issues with insights:

1. There is a legend appearing above the graph that we don't currently have
2. The graph is no longer responsive. If you look at master, the graph resizes when you dynamically change page width but this one doesn't seem to do so.

See screenshot for both issues

![Screen Shot 2022-10-14 at 8 51 16 AM](https://user-images.githubusercontent.com/14115924/195851781-d8f3e713-527b-4ddc-9731-17c4bf9db2e3.png)

m1lt0n commented:
@BenCookie95 Added a couple of changes based on your feedback. Specifically:
1. Removed the legends (thanks :) )
2. Removed the decimal points (I found that after comparing current chart in production with the one in my local)
3. Listen to resize as the width of the container is fixed between breakpoints in several occasions. In chart.js 2, there was a size monitor element created by default, which is not the case for chart.js 3, so I had to do that for responsive charts.

BenCookie95 commented:
> https://mattermost-webapp-pr-11326.test.mattermost.cloud

Thanks @m1lt0n, changes look to be working. One more small thing is that the cursor is no longer a pointer in the graph, would be great if we could have that back

BenCookie95 commented:
> @BenCookie95 Added a couple of changes based on your feedback. Specifically:
> 
> 1. Removed the legends (thanks :) )
> 2. Removed the decimal points (I found that after comparing current chart in production with the one in my local)
> 3. Listen to resize as the width of the container is fixed between breakpoints in several occasions. In chart.js 2, there was a size monitor element created by default, which is not the case for chart.js 3, so I had to do that for responsive charts.
> 
> All the changes are in the last commit if you want to check atomically.

Thanks @m1lt0n, changes look to be working. One more small thing is that the cursor is no longer a pointer in the graph, would be great if we could have that back

m1lt0n commented:
@BenCookie95 and @M-ZubairAhmed latest changes are in and pipeline green, feel free to review :)

M-ZubairAhmed commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/254978

m1lt0n commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/256646

jgilliam17 commented:
running e2es again, last cycle failed to start 
https://automation-dashboard.vercel.app/cycles/fa67a1b2-810f-4fad-8b55-118e056bc326

jgilliam17 commented:
Thanks @m1lt0n 
Tested, looks good to merge.
Verified top channel insights and system console pages - working as before.
One issue found during testing, but it seems that it self-resolved. High Availability table was not displaying initially and was taking a long time to load, refresh caused an error (see image). After few minutes, server loaded again and the issue wasn't reproducing. Just noting here what was observed, in case high availability table section is using chart.js 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/257071-b28dcbf-onprem-ent-webapp-pr-11326-mm-ee-test:b28dcbf/mochawesome.html) - no PR related failures. 
<img width="1371" alt="Screen Shot 2022-11-01 at 9 34 57 AM" src="https://user-images.githubusercontent.com/52937121/199250716-ad3e10b7-590c-4b2c-bd31-3d4feea5a202.png">
<img width="1728" alt="Screen Shot 2022-11-01 at 9 44 15 AM" src="https://user-images.githubusercontent.com/52937121/199250718-db613194-812b-4663-a0bd-f2023572b1bf.png">

m1lt0n commented:
Thanks @jgilliam17 ! High available table doesn't use chart.js. The error seems to be coming from our CDN, so perhaps some instant glitch

jgilliam17 commented:
Thanks for checking @m1lt0n I left the server available in case logs were needed.
